### ISSUE TITLE
Race condition and leak when constructing singletons

### ISSUE THREAD

**Issue by CornedBee** [structuremap/structuremap#589]
In our application we have a service that *must not* be constructed more than once. It connects to a message queue, sends messages there, and waits for replies. If there are two of this service, they will eat each other's messages. (I realize this could be more robust.)

This service is configured as a singleton. Unfortunately this doesn't guarantee that the service is only instantiated once - if two threads request the service at the same time (i.e. the second request comes in before the object is fully constructed), it is constructed twice. (It then appears to me that only one of the two is cached and used for subsequent injections, but the other one lives on - it doesn't even get disposed!)

Here is a console application that shows the problem (using .Net Core 1.1 and StructureMap 4.5.1).

```
using StructureMap;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace StructureMapThreadSafetyTest
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var container = new Container(x =>
            {
                x.Scan(s =>
                {
                    s.TheCallingAssembly();
                    s.AddAllTypesOf<IDependency>();
                });
                x.For<IDependency>().Singleton();
            }))
            {
                Task.Run(() => AsyncMain(container)).Wait();
            }
        }

        private static async Task AsyncMain(Container container)
        {
            await Task.WhenAll(
                RetrieveObject(container),
                RetrieveObject(container));
        }

        private static async Task RetrieveObject(Container container)
        {
            await Task.Yield();
            Console.WriteLine("Requesting worker");
            var uses = container.GetInstance<UsesDependency>();
            await uses.DoComplicatedWork();
        }
    }

    public interface IDependency
    {
        int Value { get; }
    }

    public class SlowConstructingDependency : IDependency, IDisposable
    {
        private static int instanceCounter;

        private int instanceNumber;

        public int Value => instanceNumber;

        public SlowConstructingDependency()
        {
            instanceNumber = Interlocked.Increment(ref instanceCounter);
            Console.WriteLine("Constructing slow dependency {0}", instanceNumber);
            Thread.Sleep(TimeSpan.FromSeconds(1));
            Console.WriteLine("Finished constructing slow dependency {0}", instanceNumber);
        }

        public void Dispose()
        {
            Console.WriteLine("Disposing slow dependency {0}", instanceNumber);
        }
    }

    public class UsesDependency
    {
        private readonly IDependency dependency;

        public UsesDependency(IDependency dependency)
        {
            Console.WriteLine("Constructing worker");
            this.dependency = dependency;
        }

        public async Task DoComplicatedWork()
        {
            Console.WriteLine("Starting complicated work");
            var value = dependency.Value;
            await Task.Delay(TimeSpan.FromSeconds(1));
            Console.WriteLine("Finished complicated work with {0}", value);
        }
    }
}
```

Actual output:

```
Requesting worker
Requesting worker
Constructing slow dependency 1
Constructing slow dependency 2
Finished constructing slow dependency 1
Finished constructing slow dependency 2
Constructing worker
Constructing worker
Starting complicated work
Starting complicated work
Finished complicated work with 1
Finished complicated work with 1
Disposing slow dependency 1
```

Note that it constructs two instances of SlowDependency, and doesn't dispose one of them.

Expected output:

```
Requesting worker
Requesting worker
Constructing slow dependency 1
Finished constructing slow dependency 1
Constructing worker
Constructing worker
Starting complicated work
Starting complicated work
Finished complicated work with 1
Finished complicated work with 1
Disposing slow dependency 1
```

Ideally it would realize that a SlowDependency is already under construction and wait for it to finish, instead of creating a second one.

**Comment by CornedBee** [comment]
Looking at the code, I think this could be accomplished by changing the dictionary in Pipeline.LifecycleObjectCache to be a `ConcurrentDictionary<int, Lazy<object>>`, so that the dictionary entry can be added immediately, before constructing the entire object. `Lazy` would then take care of thread synchronization during the construction. See this SO question: https://stackoverflow.com/questions/12611167/why-does-concurrentdictionary-getoraddkey-valuefactory-allow-the-valuefactory

Of course, this is a slight memory hit (every cache entry grows by the `Lazy` overhead, which appears to be three references in the class itself, a tiny wrapper object once the value is realized, and a sentinel object and a `Func` before), a moderate speed hit on initial access (constructing the `Lazy`, locking a mutex) a slight speed hit on subsequent access (one additional redirection, a downcast and two branches).

**Comment by udlose** [comment]
Has there been any consideration of this bug?

**Comment by jeremydmiller** [comment]
No, interested in doing the PR for it? Shouldn't be that bad to just take his suggestion up above. I'd lost track of this one.

**Comment by CornedBee** [comment]
I have implemented the change, but I need advice on where and how to test it.

**Comment by CornedBee** [comment]
Lazy's exception caching makes this not work as I want it. Will have to think further on this.

**Comment by CornedBee** [comment]
My second attempt (sort of reimplementing the relevant parts of Lazy) failed too. Putting something into the map before it is constructed causes issues with cyclic dependencies. I would have to move the cycle detection to this code.

**Comment by sm-g** [comment]
Why this still not fixed?

**Comment by jeremydmiller** [comment]
@sm-g The fix was in 4.6.0, the issue just didn't get closed.
